networking.go

```diff
@@ -133,6 +136,48 @@ func startStatsdTCP(s *Server, addr *net.TCPAddr, packetPool *sync.Pool) net.Add
 	return listener.Addr()
 }
 
+// startStatsdUnix starts listening for datagram statsd metric packets
+// on a UNIX domain socket address. It does so until the
+// server's shutdown socket is closed. startStatsdUnix returns a channel
+// that is closed once the listening connection has terminated.
+func startStatsdUnix(s *Server, addr *net.UnixAddr, packetPool *sync.Pool) (<-chan struct{}, net.Addr) {
+	done := make(chan struct{})
+	if addr.Network() != "unixgram" {
+		panic(fmt.Sprintf("Can't listen for statsd metrics on %v: only udp:// and unixgram:// addresses are supported", addr))
+	}
+	// ensure we are the only ones locking this socket:
+	lock := acquireLockForSocket(addr)
+
+	conn, err := net.ListenUnixgram(addr.Network(), addr)
```

Comment:
If I understand this code correctly, we're using the unixgram socket as a connection-less `net.Conn`; that means we're reading the datagrams in sequence with `s.ReadStatsdDatagramSocket` down below. I think it would be cool if we could parallelize reading these datagrams. 

Unfortunately, we can't use `SO_REUSEPORT` with UNIX domain sockets (so refactoring `startProcessingOnUDP` is out), but we can have multiple goroutines read off the UnixConn, which should at least cut down on the amount of sequential byte-reading/parsing we do in this loop.

Reply:
Yeah sure parallelized the metric packet reading, splitting and handling all into seperate go routine.

Reply:
Ah! What I was thinking of with that suggestion was to spawn n `ReadStatsdDatagramSocket` goroutines that read datagrams in parallel, instead of spawning new goroutines for each metric/metric batch - on this code path, we tend to prefer having fixed-size worker/reader pools rather than spawning new goroutines!

Reply:
To determine `n`, I think you can use the `num_readers` option here: https://github.com/stripe/veneur/blob/307cfea8a3b5a26858d83a0cee9861ce5a6acf2f/example.yaml#L153 - spawn this many `ReadStatsdDatagramSocket`'es and that work & read stuff in parallel! (Conveniently, datagram reads are atomic)